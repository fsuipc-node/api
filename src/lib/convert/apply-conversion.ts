import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
