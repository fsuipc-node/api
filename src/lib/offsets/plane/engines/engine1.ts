import { Type } from 'fsuipc';

import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';
import { OffsetCategory } from '@shared/offset-category';

export const engine1: OffsetList = {
  engine1ThrottleLever: new Offset({
    value: 0x88C,
    name: 'engine1ThrottleLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 throttle lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine1PropLever: new Offset({
    value: 0x88E,
    name: 'engine1PropLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 prop lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1MixtureLever: new Offset({
    value: 0x890,
    name: 'engine1MixtureLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 prop lever - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1StarterSwitchPosition: new Offset({
    value: 0x892,
    name: 'engine1StarterSwitchPosition',
    category: OffsetCategory.ENGINE,
    description: 'engine1 start switch position - JET: 0=off 1=start 2=gen/alt - PROP: 0=off 1=right 2=left 3=both 4=start',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine1Firing: new Offset({
    value: 0x894,
    name: 'engine1Firing',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 combustion',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1N2: new Offset({
    value: 0x896,
    name: 'engine1N2',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 N2 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1N1: new Offset({
    value: 0x898,
    name: 'engine1N1',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 N1 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1PropRPM: new Offset({
    value: 0x898,
    name: 'engine1PropRPM',
    category: OffsetCategory.ENGINE,
    description: 'derive RPM by multiplying by  - negative = counter-rotating propeller',
    convert: '{VAL} / 16384 / 65536',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1ThrottleLeverControl: new Offset({
    value: 0x89A,
    name: 'engine1ThrottleLeverControl',
    category: OffsetCategory.ENGINE,
    description: 'engine1 throttle lever control - -4096 to +16384',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1FuelFlowLbHourSSL: new Offset({
    value: 0x8A0,
    name: 'engine1FuelFlowLbHourSSL',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel flow (lb per hour, sea level)',
    convert: '{VAL} / 128',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1AntiIce: new Offset({
    value: 0x8B2,
    name: 'engine1AntiIce',
    category: OffsetCategory.ENGINE,
    description: 'engine1 anti-ice/carb heat active',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1OilTemp: new Offset({
    value: 0x8B8,
    name: 'engine1OilTemp',
    category: OffsetCategory.ENGINE,
    description: 'engine1 oil temp - celsius',
    convert: 'Math.round({VAL} * 140 / 16384)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine1OilPres: new Offset({
    value: 0x8BA,
    name: 'engine1OilPres',
    category: OffsetCategory.ENGINE,
    description: 'engine1 oil pressure - PSI',
    convert: 'Math.round({VAL} * 55 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1PressureRatio: new Offset({
    value: 0x8BC,
    name: 'engine1PressureRatio',
    category: OffsetCategory.ENGINE,
    description: 'engine1 pressure ratio',
    convert: '{VAL} * 1.6 / 16384',
    type: Type.Int16,
    permission: 'r',
  }),
  engine1EGT: new Offset({
    value: 0x8BE,
    name: 'engine1EGT',
    category: OffsetCategory.ENGINE,
    description: 'engine1 exhaust gas temperature - Trust only on jet engine - celsius',
    convert: 'Math.round({VAL} * 860 / 16384)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine1MP: new Offset({
    value: 0x8C0,
    name: 'engine1MP',
    category: OffsetCategory.ENGINE,
    description: 'engine1 manifold pressure - inHg',
    convert: '{VAL} / 1024',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1RPMScaler: new Offset({
    value: 0x8C8,
    name: 'engine1RPMScaler',
    category: OffsetCategory.ENGINE,
    description: 'engine1 RPM Scaler - props: used to calculate RPM',
    type: Type.Int16,
    permission: 'r',
  }),
  engine1OilQuantity: new Offset({
    value: 0x8D0,
    name: 'engine1OilQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine1 oil quantity - percent',
    convert: 'Math.round({VAL} / 16384 * 100)',
    type: Type.Int32,
    permission: 'r',
  }),
  engine1Vibration: new Offset({
    value: 0x8D4,
    name: 'engine1Vibration',
    category: OffsetCategory.ENGINE,
    description: 'engine1 vibration',
    convert: 'Math.round({VAL} * 5 / 16384)',
    type: Type.Int32,
    permission: 'r',
  }),
  engine1HydPres: new Offset({
    value: 0x8D8,
    name: 'engine1HydPres',
    category: OffsetCategory.ENGINE,
    description: 'engine1 hydraulic pressure - PSI',
    convert: '{VAL} / 4',
    type: Type.Int32,
    permission: 'r',
  }),
  engine1HydQuantity: new Offset({
    value: 0x8DC,
    name: 'engine1HydQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine1 hydraulic quantity - PSI',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int32,
    permission: 'r',
  }),
  engine1CHT: new Offset({
    value: 0x8E8,
    name: 'engine1CHT',
    category: OffsetCategory.ENGINE,
    description: 'engine1 cylinder head temperature - F',
    type: Type.Double,
    permission: 'r',
  }),
  engine1ITT: new Offset({
    value: 0x8F0,
    name: 'engine1ITT',
    category: OffsetCategory.ENGINE,
    description: 'engine1 turbine temperature - C',
    convert: '{VAL} / 16384',
    type: Type.Int32,
    permission: 'r',
  }),
  engine1Torque: new Offset({
    value: 0x8F4,
    name: 'engine1Torque',
    category: OffsetCategory.ENGINE,
    description: 'engine1 torque - trust only for helo - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int32,
    permission: 'r',
  }),
  engine1FuelPres: new Offset({
    value: 0x8F8,
    name: 'engine1FuelPres',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel pressure - PSI',
    convert: '{VAL} / 144',
    type: Type.Int32,
    permission: 'r',
  }),
  engine1FuelUsedSinceStart: new Offset({
    value: 0x90C,
    name: 'engine1FuelUsedSinceStart',
    category: OffsetCategory.ENGINE,
    description: 'engine1 used fuel since start - lb',
    type: Type.Single,
    permission: 'r',
  }),
  engine1FuelElapsedTime: new Offset({
    value: 0x910,
    name: 'engine1FuelElapsedTime',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel elapsed time - h',
    type: Type.Single,
    permission: 'r',
  }),
  engine1FuelFlowLbHour: new Offset({
    value: 0x918,
    name: 'engine1FuelFlowLbHour',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel elapsed time - lb/h',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TorqueFtLb: new Offset({
    value: 0x920,
    name: 'engine1TorqueFtLb',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel elapsed time - ft lb',
    type: Type.Single,
    permission: 'r',
  }),
  engine1TurbineN1: new Offset({
    value: 0x2000,
    name: 'engine1TurbineN1',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine N1 - percent',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineN2: new Offset({
    value: 0x2008,
    name: 'engine1TurbineN2',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine N2 - percent',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineCorrectedN1: new Offset({
    value: 0x2010,
    name: 'engine1TurbineCorrectedN1',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine corrected N1 - percent',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineCorrectedN2: new Offset({
    value: 0x2018,
    name: 'engine1TurbineCorrectedN2',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine corrected N2 - percent',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineCorrectedFuelFlow: new Offset({
    value: 0x2020,
    name: 'engine1TurbineCorrectedFuelFlow',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine corrected fuel flow - lb/h',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineTorque: new Offset({
    value: 0x2028,
    name: 'engine1TurbineTorque',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine corrected fuel flow - 0.0 to 1.0',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineEPR: new Offset({
    value: 0x2030,
    name: 'engine1TurbineEPR',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine EPR',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineITT: new Offset({
    value: 0x2038,
    name: 'engine1TurbineITT',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine ITT',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineAfterburnerActive: new Offset({
    value: 0x2048,
    name: 'engine1TurbineAfterburnerActive',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine afterburner active',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1TurbineJetThrust: new Offset({
    value: 0x204C,
    name: 'engine1TurbineJetThrust',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine jet thrust',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineTankSelector: new Offset({
    value: 0x2054,
    name: 'engine1TurbineTankSelector',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine tank selector',
    convert: 'fuelTank',
    mapping: true,
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1TurbineUsedTank: new Offset({
    value: 0x2058,
    name: 'engine1TurbineUsedTank',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine used tank',
    type: Type.BitArray,
    length: 11,
    permission: 'r',
  }),
  engine1TurbineAvailableTankCount: new Offset({
    value: 0x205C,
    name: 'engine1TurbineAvailableTankCount',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine available tank count',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1TurbineFuelFlow: new Offset({
    value: 0x2060,
    name: 'engine1TurbineFuelFlow',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine fuel flow',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineIsFuelAvailable: new Offset({
    value: 0x2068,
    name: 'engine1TurbineIsFuelAvailable',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine is fuel available flag',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1TurbineBleedPressure: new Offset({
    value: 0x206C,
    name: 'engine1TurbineBleedPressure',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine bleed air pressure - lb/sq in',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineReverser: new Offset({
    value: 0x207C,
    name: 'engine1TurbineReverser',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine reverser position - 0.0 - 1.0 lb/sq in',
    convert: 'Math.round(+(+({VAL}).toFixed(4) * 100).toFixed(2))',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineVibration: new Offset({
    value: 0x2084,
    name: 'engine1TurbineVibration',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine vibration',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TurbineIgnition: new Offset({
    value: 0x208C,
    name: 'engine1TurbineIgnition',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 turbine is fuel available flag',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
};
