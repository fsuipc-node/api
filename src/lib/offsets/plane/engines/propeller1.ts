import { Type } from 'fsuipc';

import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';
import { OffsetCategory } from '@shared/offset-category';

export const propeller1: OffsetList = {
  propeller1RPM: new Offset({
    value: 0x2400,
    name: 'propeller1RPM',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 RPM',
    type: Type.Double,
    permission: 'r',
  }),
  propeller1RPMPercent: new Offset({
    value: 0x2408,
    name: 'propeller1RPMPercent',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 RPM percent',
    convert: 'Math.round(+(+({VAL}).toFixed(4) * 100).toFixed(2))',
    type: Type.Double,
    permission: 'r',
  }),
  propeller1Thrust: new Offset({
    value: 0x2410,
    name: 'propeller1Thrust',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 thrust - lb',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Double,
    permission: 'r',
  }),
  propeller1BetaBladeAngle: new Offset({
    value: 0x2418,
    name: 'propeller1BetaBladeAngle',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 beta blade angle - rad',
    convert: '+({VAL}).toFixed(4)',
    type: Type.Double,
    permission: 'r',
  }),
  propeller1FeatheringInhibit: new Offset({
    value: 0x2420,
    name: 'propeller1FeatheringInhibit',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 feathering inhibit',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller1Feathered: new Offset({
    value: 0x2424,
    name: 'propeller1Feathered',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 feathered',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  propeller1SyncDeltaLever: new Offset({
    value: 0x2428,
    name: 'propeller1SyncDeltaLever',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 sync delta lever',
    type: Type.Double,
    permission: 'r',
  }),
  propeller1AutofeatherArmed: new Offset({
    value: 0x2430,
    name: 'propeller1AutofeatherArmed',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 autofeather armed flag',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller1FeatherSwitch: new Offset({
    value: 0x2434,
    name: 'propeller1FeatherSwitch',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 feather switch',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller1PanelAutofeatherSwitch: new Offset({
    value: 0x2438,
    name: 'propeller1PanelAutofeatherSwitch',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 panel autofeather switch',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller1Sync: new Offset({
    value: 0x243C,
    name: 'propeller1Sync',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 sync active',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller1DeiceSwitch: new Offset({
    value: 0x2440,
    name: 'propeller1DeiceSwitch',
    category: OffsetCategory.ENGINE,
    description: 'propeller 1 de-ice active',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
};
