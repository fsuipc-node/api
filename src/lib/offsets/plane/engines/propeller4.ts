import { Type } from 'fsuipc';

import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';
import { OffsetCategory } from '@shared/offset-category';

export const propeller4: OffsetList = {
  propeller4RPM: new Offset({
    value: 0x2700,
    name: 'propeller4RPM',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 RPM',
    type: Type.Double,
    permission: 'r',
  }),
  propeller4RPMPercent: new Offset({
    value: 0x2708,
    name: 'propeller4RPMPercent',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 RPM percent',
    convert: 'Math.round(+(+({VAL}).toFixed(4) * 100).toFixed(2))',
    type: Type.Double,
    permission: 'r',
  }),
  propeller4Thrust: new Offset({
    value: 0x2710,
    name: 'propeller4Thrust',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 thrust - lb',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Double,
    permission: 'r',
  }),
  propeller4BetaBladeAngle: new Offset({
    value: 0x2718,
    name: 'propeller4BetaBladeAngle',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 beta blade angle - rad',
    convert: '+({VAL}).toFixed(4)',
    type: Type.Double,
    permission: 'r',
  }),
  propeller4FeatheringInhibit: new Offset({
    value: 0x2720,
    name: 'propeller4FeatheringInhibit',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 feathering inhibit',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller4Feathered: new Offset({
    value: 0x2724,
    name: 'propeller4Feathered',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 feathered',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  propeller4SyncDeltaLever: new Offset({
    value: 0x2728,
    name: 'propeller4SyncDeltaLever',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 sync delta lever',
    type: Type.Double,
    permission: 'r',
  }),
  propeller4AutofeatherArmed: new Offset({
    value: 0x2730,
    name: 'propeller4AutofeatherArmed',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 autofeather armed flag',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller4FeatherSwitch: new Offset({
    value: 0x2734,
    name: 'propeller4FeatherSwitch',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 feather switch',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller4PanelAutofeatherSwitch: new Offset({
    value: 0x2738,
    name: 'propeller4PanelAutofeatherSwitch',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 panel autofeather switch',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller4Sync: new Offset({
    value: 0x273C,
    name: 'propeller4Sync',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 sync active',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propeller4DeiceSwitch: new Offset({
    value: 0x2740,
    name: 'propeller4DeiceSwitch',
    category: OffsetCategory.ENGINE,
    description: 'propeller 4 de-ice active',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
};
