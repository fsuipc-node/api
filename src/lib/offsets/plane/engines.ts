import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

import { engine1 } from './engines/engine1';
import { engine2 } from './engines/engine2';
import { engine3 } from './engines/engine3';
import { engine4 } from './engines/engine4';

export const engines: OffsetList = {
  engineType: new Offset({
    value: 0x609,
    name: 'engineType',
    category: OffsetCategory.ENGINE,
    description: 'engine type',
    convert: 'engineType',
    mapping: true,
    type: Type.Byte,
    permission: 'r',
  }),
  hasMixtureControl: new Offset({
    value: 0x780,
    name: 'hasMixtureControl',
    category: OffsetCategory.ENGINE,
    description: 'has mixture control',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasCarbHeat: new Offset({
    value: 0x784,
    name: 'hasCarbHeat',
    category: OffsetCategory.ENGINE,
    description: 'has carb heat',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  activeEngine: new Offset({
    value: 0x888,
    name: 'activeEngine',
    category: OffsetCategory.ENGINE,
    description: 'active engine pattern',
    type: Type.BitArray,
    length: 1,
    permission: 'rw',
  }),
  ...engine1,
  ...engine2,
  ...engine3,
  ...engine4,
};
