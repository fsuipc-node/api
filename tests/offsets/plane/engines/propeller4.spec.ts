import { propeller4 as offsets } from '@offsets/plane/engines/propeller4';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/engines/propeller4', () => {
  const offsetsTestCases = [
    { name: 'propeller4RPMPercent', value: 0.02386, expectedResult: 2 },
    { name: 'propeller4Thrust', value: 23.8601932, expectedResult: 23.86 },
    { name: 'propeller4BetaBladeAngle', value: 23.8601932, expectedResult: 23.8602 },
    { name: 'propeller4FeatheringInhibit', value: 0, expectedResult: false },
    { name: 'propeller4FeatheringInhibit', value: 1, expectedResult: true },
    { name: 'propeller4Feathered', value: 0, expectedResult: false },
    { name: 'propeller4Feathered', value: 1, expectedResult: true },
    { name: 'propeller4AutofeatherArmed', value: 0, expectedResult: false },
    { name: 'propeller4AutofeatherArmed', value: 1, expectedResult: true },
    { name: 'propeller4FeatherSwitch', value: 0, expectedResult: false },
    { name: 'propeller4FeatherSwitch', value: 1, expectedResult: true },
    { name: 'propeller4PanelAutofeatherSwitch', value: 0, expectedResult: false },
    { name: 'propeller4PanelAutofeatherSwitch', value: 1, expectedResult: true },
    { name: 'propeller4Sync', value: 0, expectedResult: false },
    { name: 'propeller4Sync', value: 1, expectedResult: true },
    { name: 'propeller4DeiceSwitch', value: 0, expectedResult: false },
    { name: 'propeller4DeiceSwitch', value: 1, expectedResult: true },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
