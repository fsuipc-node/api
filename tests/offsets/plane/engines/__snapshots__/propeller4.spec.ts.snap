// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/engines/propeller4 offsets list should have required properties 1`] = `
Object {
  "propeller4AutofeatherArmed": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "propeller 4 autofeather armed flag",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4AutofeatherArmed",
    "permission": "rw",
    "type": 6,
    "value": 10032,
  },
  "propeller4BetaBladeAngle": Offset {
    "category": "engine",
    "convert": "+({VAL}).toFixed(4)",
    "description": "propeller 4 beta blade angle - rad",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4BetaBladeAngle",
    "permission": "r",
    "type": 8,
    "value": 10008,
  },
  "propeller4DeiceSwitch": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "propeller 4 de-ice active",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4DeiceSwitch",
    "permission": "rw",
    "type": 6,
    "value": 10048,
  },
  "propeller4FeatherSwitch": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "propeller 4 feather switch",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4FeatherSwitch",
    "permission": "rw",
    "type": 6,
    "value": 10036,
  },
  "propeller4Feathered": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "propeller 4 feathered",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4Feathered",
    "permission": "r",
    "type": 6,
    "value": 10020,
  },
  "propeller4FeatheringInhibit": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "propeller 4 feathering inhibit",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4FeatheringInhibit",
    "permission": "rw",
    "type": 6,
    "value": 10016,
  },
  "propeller4PanelAutofeatherSwitch": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "propeller 4 panel autofeather switch",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4PanelAutofeatherSwitch",
    "permission": "rw",
    "type": 6,
    "value": 10040,
  },
  "propeller4RPM": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "propeller 4 RPM",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4RPM",
    "permission": "r",
    "type": 8,
    "value": 9984,
  },
  "propeller4RPMPercent": Offset {
    "category": "engine",
    "convert": "Math.round(+(+({VAL}).toFixed(4) * 100).toFixed(2))",
    "description": "propeller 4 RPM percent",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4RPMPercent",
    "permission": "r",
    "type": 8,
    "value": 9992,
  },
  "propeller4Sync": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "propeller 4 sync active",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4Sync",
    "permission": "rw",
    "type": 6,
    "value": 10044,
  },
  "propeller4SyncDeltaLever": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "propeller 4 sync delta lever",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4SyncDeltaLever",
    "permission": "r",
    "type": 8,
    "value": 10024,
  },
  "propeller4Thrust": Offset {
    "category": "engine",
    "convert": "+({VAL}).toFixed(2)",
    "description": "propeller 4 thrust - lb",
    "length": undefined,
    "mapping": undefined,
    "name": "propeller4Thrust",
    "permission": "r",
    "type": 8,
    "value": 10000,
  },
}
`;
